File: DesignAParkingLot.py
-------------------------------Design a parking lot ---------------------------------------------
# Time Complexity : park: O(1), unpark o(log(floor*spots))
# Space Complexity : O(floors*spots)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this : No
# 
# Here we will first design a parking lot with number of floors and and spots for each floor.
#and we will have 2 functions which are park and unpark. We will first have vacant spots 
# and all the spots will be added to the heap which sorts by the ascending order or floors and then spots.
# Once the spot is filled up we will pop the spot from the heap. When it is vacant we will add to the heap again .
# so that we will find it in the order of floor and spot number.

import heapq
class HeapClass:
  def __init__(self, floor, spot):
    self.floor = floor
    self.spot = spot
  
  def __lt__(self, other):
    if self.floor == other.floor:
      return self.spot<other.spot
    else:
      return self.floor<other.floor

class Parkinglot:
  def __init__(self,maxFloors, maxSpots):
    self.maxFloors = maxFloors
    self.maxSpots = maxSpots
    self.heap = []
  
  def park(self):
    lot = heapq.heappop(self.heap)
    print(lot.floor, lot.spot)
    return lot
  
  def unpark(self, floor, spot):

    heapq.heappush(self.heap, HeapClass(floor, spot))
  
pl = Parkinglot(2,4)

for i in range(pl.maxFloors):
  for j in range(pl.maxSpots):
    heapq.heappush(pl.heap, HeapClass(i,j))

print((pl.heap[0].floor, pl.heap[0].spot))
pl.park()
pl.park()
pl.park()
pl.park()
print((pl.heap[0].floor, pl.heap[0].spot))
pl.unpark(0,2)
print((pl.heap[0].floor, pl.heap[0].spot))

File: Design a parking lot.java
class Main {
    
    static class ParkingSpot
    {
        int spot;
        int floor;
        
        public ParkingSpot(int floor,int spot)
        {
            this.floor=floor;
            this.spot=spot;
        }   
        public int getFloor()
        {
            return this.floor;
        }
        public int getSpot()
        {
            return this.spot;
        }
        
    }
    
    static class ParkingLot
    {
        int maxFloor;
        int maxSpot;
        PriorityQueue <ParkingSpot> pq=new PriorityQueue<>((a,b)->
                                             {if(a.floor==b.floor) return a.spot-b.spot;
                                            return a.floor-b.floor;
                                             });
          
        public ParkingLot(int maxFloor,int maxSpot)
        {
            this.maxFloor=maxFloor;
            this.maxSpot=maxSpot;
        }
        // park
        public ParkingSpot park()
        {
            //if there are no vacant spots in the priority queue
            if(pq.isEmpty())
            {
                throw new IllegalStateException("Parking lot is full") ;
            }
            ParkingSpot re=pq.remove();
            return re;
        }
        
        //unpark
        public void unpark(int floor,int spot)
        {
            ParkingSpot newspot=new ParkingSpot(floor,spot);
             pq.add(newspot);
        }
        
        //get the next available spot
        public ParkingSpot getNextAvailableSpot()
        {
            return pq.peek();
        }
        //add new spot
        public void addParkingSpot(int floor, int spot)
        {
            if(floor>maxFloor)
            {
                throw new IllegalStateException("floor input greater than max allowed"); 
            }
            if(spot>maxSpot)
            {
                throw new IllegalStateException("spot input greater than max allowed"); 
            }
            
            ParkingSpot newSpot=new ParkingSpot(floor, spot);
            pq.add(newSpot);
        }
    }
    
    
    
    public static void main(String[] args) {
        ParkingLot p1=new ParkingLot(10,20); //maxFloor=10 & max spots in each floor=20
            p1.addParkingSpot(1, 1);
            p1.addParkingSpot(2, 1);
            p1.addParkingSpot(3, 1);
            p1.addParkingSpot(1, 2);
            p1.addParkingSpot(2, 2);
            p1.addParkingSpot(3, 2);
            
            ParkingSpot n=p1.getNextAvailableSpot();
        System.out.println("Parked at floor" + n.getFloor() + " ,Spot=" + n.getSpot());
            p1.park();
            ParkingSpot n1=p1.getNextAvailableSpot();
        System.out.println("Parked at floor" + n1.getFloor() + " ,Spot=" + n1.getSpot());
            
            p1.unpark(1,1);
            ParkingSpot n2=p1.getNextAvailableSpot();
        System.out.println("Parked at floor" + n2.getFloor() + " ,Spot=" + n2.getSpot());
    }
}


      
  
    